validator/src/journal/chain.rs

```diff
@@ -531,6 +531,63 @@ impl<
         }
     }
 
+    // Returns all blocks in forks not on the chain with the given head. If head is None, uses the
+    // current chain head. If head is not found, returns None.
+    pub fn forks(&self, head: &str) -> Option<Vec<Block>> {
+        let state = self
+            .state
+            .read()
+            .expect("No lock holder should have poisoned the lock");
+
+        match state.block_manager.ref_block(head) {
+            Err(BlockManagerError::UnknownBlock) => {
+                return None;
+            }
+            Err(err) => {
+                error!("Unexpected error occurred: {:?}", err);
+                return None;
+            }
+            Ok(_) => (),
+        }
+
+        let mut forks: Vec<Option<Block>> = state
+            .fork_cache
+            .forks()
+            .into_iter()
+            .flat_map(|fork_head: &String| {
+                state
+                    .block_manager
+                    .branch_diff(fork_head, head)
+                    .expect("Fork not found, but should be referenced")
+                    .map(|block| Some(block))
+            }).collect();
+
+        for i in 0..forks.len() {
```

Comment:
It's removing any duplicates from the list. I'll move it behind a function. I can't use the `dedup()` method because I need to preserve the original order.

Reply:
Why does order matter, after you've flattened them?

Reply:
The contract between the validator and consensus engine is that the validator will send new blocks to the engine consecutively and increasing for the block height, on a given fork. I had intended this function to return blocks in an order that was consistent with that contract, but now that you are questioning it, I see that I actually got the order reversed. Thinking about it more though, if I sort the list in increasing order of block height, I think it still satisfies that contract and I can just use dedup then.